### PR TITLE
avoid_list now supports <s> in negative constraints

### PR DESCRIPTION
**Pull request by mjpost**
Simple fix that lets you add `<s>` as the beginning of a sequence to avoid.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: contrib/sacrebleu/sacrebleu.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
 import math
 import unicodedata
 
-VERSION = '1.3.0'
+VERSION = '1.2.10'
```
Comment (fhieber) [inline contrib/sacrebleu/sacrebleu.py]: not sure if sneaking this in is the right way. Why not make a small separate PR and explain the downgrade?

File: sockeye/inference.py [hunk 2]
```diff
@@ -1461,6 +1461,7 @@ def _beam_search(self,
             avoid_states = constrained.AvoidBatch(self.batch_size, self.beam_size,
                                                   avoid_list=raw_avoid_list,
                                                   global_avoid_trie=self.global_avoid_trie)
+            avoid_states.consume(best_word_indices)
```
Comment (fhieber) [inline sockeye/inference.py]: any chance for a test? :)

**Comment by tdomhan** [comment]
what does it mean to allow to avoid `<s>`?

**Comment by tdomhan** [comment]
is the idea to have an avoid phrase to only apply at the beginning of the sentence?

**Comment by fhieber** [comment]
I would think that this enables users to say that the translation should not start with "I think", when specifying an avoid constraint as such: "<s> I think".

File: sockeye/inference.py (hunk 2 above)
Comment (mjpost) [reply sockeye/inference.py]: I can't sneak anything by you! Okay, I'll add one.

**Comment by mjpost** [comment]
Yes, exactly. This allows you to avoid a prefix, e.g., "<s> I think". Trying to specify "<s>" has no effect.

**Comment by tdomhan** [comment]
Thanks! Could you add that explanation to the CHANGELOG?

**Comment by mjpost** [comment]
Okay, I added test cases and refactored them in the process. Adding the test cases uncovered a bug.

This had to be an integration test since it is testing the output of `_beam_search()`. But the sequence copy integration test was already getting unwieldy, and was never ideal for testing constraints. The refactored tests (a) remove testing constraints from `test_seq_copy_int.py` and move it into `test_constraints_int.py`. There, we first decode in an unconstrained setting, then add positive constraints, and negative constraints, and ensure that the proper thing happens in the output.

This uncovered a bug. Say you have a negative constraint of "6 5 5". This wouldn't prevent the output "6 6 5 5" from being generated, because the state wasn't getting properly reset when generating the second "6". The code was resetting the state to the beginning of the trie, but actually, it needed to be taking the first step from the root of the trie.

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,12 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
 Each version section may have have subsections for: _Added_, _Changed_, _Removed_, _Deprecated_, and _Fixed_.
 
+## [1.18.43]
+### Fixed
+- `<s>` now supported as the first token in a multi-word negative constraint
```
Comment (fhieber) [inline CHANGELOG.md]: [nit] could make this a bullet point and the item in 1.18.42 as well?

File: sockeye/lexical_constraints.py [hunk 4]
```diff
@@ -116,10 +116,21 @@ def consume(self, word_id: int) -> 'AvoidState':
         """
         Consumes a word, and updates the state based on it. Returns new objects on a state change.
 
+        The next state for a word can be tricky. Here are the cases:
+        (1) If the word is found in our set of outgoing child arcs, we take that transition.
+        (2) If the word is not found, and we are not in the root state, we need to reset.
+            This means we pretend we were in the root state, and see if we can take a step
+        (3) Otherwise, if we are not already the root state (i.e., we were partially through
```
Comment (fhieber) [inline sockeye/lexical_constraints.py]: "if we are not already IN the root state"

File: sockeye/lexical_constraints.py [hunk 5]
```diff
@@ -116,10 +116,21 @@ def consume(self, word_id: int) -> 'AvoidState':
...
+            the trie), we need to create a new object to track the state, in the root state
```
Comment (fhieber) [inline sockeye/lexical_constraints.py]: "create a new object in root state" maybe?

File: test/common.py [hunk 6]
```diff
@@ -375,6 +347,61 @@ def run_train_translate(train_params: str,
...
+                                                        _TRANSLATE_PARAMS_COMMON.format(model=model_path,
+                                                                                        input=new_test_source_path,
+                                                                                        output=out_path_constrained,
+                                                                                        quiet=quiet_arg),
+                                                        translate_params)
+
+                with patch.object(sys, "argv", params.split()):
+                    sockeye.translate.main()
+
+                print('JSON_INPUT', open(new_test_source_path).readlines(),
```
Comment (fhieber) [inline test/common.py]: could we change this to a logging statement? with pytest, logging statements are easier to track than stdout output.

File: test/common.py [hunk 7]
```diff
@@ -375,6 +347,61 @@ def run_train_translate(train_params: str,
...
+                      'CONSTRAINED', open(out_path_constrained).readlines(),
+                      'UNCONSTRAINED', open(out_path).readlines(), sep='\n')
+                for json_input, constrained_out, unconstrained_out in zip(open(new_test_source_path), open(out_path_constrained), open(out_path)):
+                    jobj = json.loads(json_input)
+                    print('JSON_INPUT', json_input, 'CONSTRAINED', constrained_out, 'UNCONSTRAINED', unconstrained_out, sep='\n')
```
Comment (fhieber) [inline test/common.py]: same here

File: test/common.py [hunk 8]
```diff
@@ -375,6 +347,61 @@ def run_train_translate(train_params: str,
...
+                    if jobj.get(constraint_type, None) == None:
+                        assert constrained_out == unconstrained_out
```
Comment (fhieber) [inline test/common.py]: would it be helpful to add error messages to these `assert` statements?

File: test/common.py (hunk 6 above)
Comment (mjpost) [reply test/common.py]: Oh, this should be removed entirely.

File: test/common.py (hunk 8 above)
Comment (mjpost) [reply test/common.py]: I can add comments, but none of the other asserts have comments, and I'm not sure exactly what to put...

File: CHANGELOG.md (hunk 3 above)
Comment (mjpost) [reply CHANGELOG.md]: Sorry, I'm not sure I understand this one. This improvement doesn't apply to 1.18.42, right?

File: test/common.py (hunk 8 above)
Comment (fhieber) [reply test/common.py]: lets leave as is, was just wondering about whether the if cases are clear, and/or what we are asserting. but after a second look its clear to me.

File: CHANGELOG.md (hunk 3 above)
Comment (fhieber) [reply CHANGELOG.md]: I should be more clear: the description is missing a dash "-", so does the previous entry in the change log (length calculation).

**Comment by mjpost** [comment]
Running system tests, shouldn't be affected, will wait to merge till they finish.

File: CHANGELOG.md (hunk 3 above)
Comment (mjpost) [reply CHANGELOG.md]: Where is the dash missing? This is what I see right now:

```
## [1.18.43]
### Added
- `<s>` now supported as the first token in a multi-word negative constraint
  (e.g., '<s> I think' to prevent a sentence from starting with 'I think')
### Fixed
- Bugfix in resetting the state of a multiple-word negative constraint

## [1.18.42]
### Changed
- Simplified gluon blocks for length calculation
```

**Comment by mjpost** [comment]
System tests pass.
